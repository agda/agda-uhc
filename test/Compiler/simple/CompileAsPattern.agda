{-# OPTIONS -v treeless.opt:20 #-}
module _ where

open import Common.Prelude

module _ {a} {A : Set a} (_<_ : A → A → Bool) where
  merge : List A → List A → List A
  merge [] ys = ys
  merge xs [] = xs
  merge xs@(x ∷ xs₁) ys@(y ∷ ys₁) =
    if x < y then (x ∷ merge xs₁ ys)  -- Generated treeless code should preserve
             else (y ∷ merge xs ys₁)  -- xs and ys (and not expand to _∷_).

open import Agda.Builtin.Nat

mapM! : {A : Set} → (A → IO Unit) → List A → IO Unit
mapM! f [] = return unit
mapM! f (x ∷ xs) = f x >>= λ _ → mapM! f xs

xs ys : List Nat
xs = 2 ∷ 3 ∷ 5 ∷ 10 ∷ 20 ∷ []
ys = 0 ∷ 1 ∷ 2 ∷ 8 ∷ 10 ∷ 15 ∷ []

main : IO Unit
main = mapM! printNat (merge _<_ xs ys)
