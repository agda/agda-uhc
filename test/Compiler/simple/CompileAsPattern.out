EXECUTED_PROGRAM

ret > ExitSuccess
out > CompileAsPattern._.merge =
out >   λ a b c d e →
out >     case d of
out >       Agda.Builtin.List.List.[] → e
out >       Agda.Builtin.List.List._∷_ f g →
out >         case e of
out >           Agda.Builtin.List.List.[] → d
out >           Agda.Builtin.List.List._∷_ h i →
out >             Common.Bool.if_then_else_
out >               _ _ (c f h)
out >               (Agda.Builtin.List.List._∷_ f (CompileAsPattern._.merge a _ c g e))
out >               (Agda.Builtin.List.List._∷_ h (CompileAsPattern._.merge a _ c d i))
out > CompileAsPattern.mapM! =
out >   λ a b c →
out >     case c of
out >       Agda.Builtin.List.List.[] → Common.IO.return () _ _
out >       Agda.Builtin.List.List._∷_ d e →
out >         Common.IO._>>=_
out >           () () _ _ (b d) (λ f → CompileAsPattern.mapM! _ b e)
out > CompileAsPattern.xs =
out >   Agda.Builtin.List.List._∷_
out >     2
out >     (Agda.Builtin.List.List._∷_
out >        3
out >        (Agda.Builtin.List.List._∷_
out >           5
out >           (Agda.Builtin.List.List._∷_
out >              10 (Agda.Builtin.List.List._∷_ 20 Agda.Builtin.List.List.[]))))
out > CompileAsPattern.ys =
out >   Agda.Builtin.List.List._∷_
out >     0
out >     (Agda.Builtin.List.List._∷_
out >        1
out >        (Agda.Builtin.List.List._∷_
out >           2
out >           (Agda.Builtin.List.List._∷_
out >              8
out >              (Agda.Builtin.List.List._∷_
out >                 10 (Agda.Builtin.List.List._∷_ 15 Agda.Builtin.List.List.[])))))
out > CompileAsPattern.main =
out >   CompileAsPattern.mapM!
out >     _ Common.IO.printNat
out >     (CompileAsPattern._.merge
out >        _ _ _<_ CompileAsPattern.xs CompileAsPattern.ys)
out > 012235810101520
